### PR TITLE
Extension serve local asset hot reload

### PR DESCRIPTION
**Pull request by ulgut**
🎩 :
1. Checkout `extension-serve-local-asset-hot-reload`
2. In an extension directory, run `shopify-dev extension serve`
3. Navigate to `localhost:9292`
4. Make some changes to the extension's asset files and note the hot reloading

### CODE REVIEW

File: lib/shopify_cli/theme/app_extension.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+require_relative "file"
+require_relative "theme_admin_api"
+
+require "pathname"
+require "time"
+
+module ShopifyCLI
+  module Theme
+    class AppExtension
```
Comment (jessetuglu) [inline lib/shopify_cli/theme/app_extension.rb:10]: @karreiro this follows the same method structure of Theme but I felt it didn't warrant a direct inheritance

File: lib/shopify_cli/theme/dev_server/app_extension_watcher.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+require "forwardable"
+
+module ShopifyCLI
+  module Theme
+    module DevServer
+      # Watches for file changes and publish events to the theme
+      class AppExtensionWatcher
+        extend Forwardable
+
+        def_delegators :@listener, :add_observer, :changed, :notify_observers
```
Comment (jessetuglu) [inline lib/shopify_cli/theme/dev_server/app_extension_watcher.rb:12]: I left these for usage in tests later, but currently only `add_observer` is being used.

File: lib/shopify_cli/theme/dev_server/local_assets_base.rb [hunk 3]
```diff
@@ -42,23 +39,11 @@ def call(env)
...
-            ]
-          else
-            fail(404, "Not found")
-          end
+        def serve_file(_path_info)
+          raise "#{self.class.name}#serve_file(path_info) must be defined"
         end
 
-        def fail(status, body)
+        def serve_fail(status, body)
```
Comment (jessetuglu) [inline lib/shopify_cli/theme/dev_server/local_assets_base.rb:46]: changed name due to Rubocop issues

**Review by karreiro — COMMENTED** [review]
Thank you, @jessetuglu. Great stuff! It's so nice to use hot-reload with assets already 🔥 



I've force-pushed this branch, because even after merging @mgmanzella PR, GitHub was still showing the commits already on `main-extension-serve` (here's the [backup of the branch](https://github.com/karreiro/shopify-cli/tree/extension-serve-local-asset-hot-reload-bkp) before the force-push, just in case I've missed something during that process - but I don't think we're missing anything).

Also, I've included an extra commit to fix a small detail in the `hot-reload.js`. When a TAE has many blocks, each one has its own stylesheet. Now, we're covering all of them :)

I'm still reviewing the PR, it works pretty nicely so far!

**Review by mgmanzella — APPROVED** [review]
noice! no blocking comments

**Review by karreiro — APPROVED** [review]
Thank you, @jessetuglu!

**Pull request merged**